### PR TITLE
wifi: nxp: Add full AP mode support (using hostapd)

### PR DESCRIPTION
**Pull request by nxf58150**
Added support of softAP setup with hostapd.

### CODE REVIEW

File: hostapd/hostapd_cli_zephyr.h [hunk 1]
```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
```
Comment (jukkar) [inline hostapd/hostapd_cli_zephyr.h:4]: wrong license

File: src/drivers/driver_zephyr.c [hunk 2]
```diff
@@ -36,6 +36,63 @@ const struct zep_wpa_supp_dev_ops *get_dev_ops(const struct device *dev)
...
+#ifdef CONFIG_WIFI_NM_HOSTAPD_AP
+void hostapd_event_wrapper(void *ctx, enum wpa_event_type event, union wpa_event_data *data)
+{
+	struct wpa_supplicant_event_msg msg = { 0 };
+
+	msg.hostapd = 1;
+	msg.ctx   = ctx;
+	msg.event = event;
+	if (data)
+	{
```
Comment (jukkar) [inline src/drivers/driver_zephyr.c:48]: please place starting `{` to previous line

File: src/drivers/driver_zephyr.c [hunk 3]
```diff
@@ -1102,6 +1182,127 @@ static void wpa_drv_zep_deinit(void *priv)
...
+	//callbk_fns.survey_res      = wpa_drv_zep_event_proc_survey_res;
+	callbk_fns.acs_channel_sel   = wpa_drv_zep_event_acs_channel_selected;
+	callbk_fns.mgmt_rx           = wpa_drv_zep_event_mgmt_rx;
+	//callbk_fns.eapol_rx        = wpa_drv_zep_event_proc_eapol_rx;
+	callbk_fns.mgmt_tx_status    = wpa_drv_zep_event_mgmt_tx_status;
+	callbk_fns.mac_changed       = wpa_drv_zep_event_mac_changed;
+	callbk_fns.chan_list_changed = wpa_drv_zep_event_chan_list_changed;
+	//callbk_fns.ecsa_complete   = wpa_drv_zep_event_ecsa_complete;
+	//callbk_fns.dfs_cac_started  = wpa_drv_zep_event_dfs_cac_started;
+	//callbk_fns.dfs_cac_finished = wpa_drv_zep_event_dfs_cac_finished;
```
Comment (jukkar) [inline src/drivers/driver_zephyr.c]: Why these commented lines, is this code still WIP and you expect to uncomment these soonish?

File: src/drivers/driver_zephyr.c [hunk 4]
```diff
@@ -905,6 +905,16 @@ static void wpa_drv_zep_event_mgmt_rx(struct zep_drv_if_ctx *if_ctx,
 	wpa_supplicant_event_wrapper(if_ctx->supp_if_ctx, EVENT_RX_MGMT, &event);
```
Comment (jukkar) [inline src/drivers/driver_zephyr.c:914]: Please do not leave empty commit messages but write a proper text there, it is not enough to just have commit subject.

File: hostapd/hostapd_cli_cmds.c [hunk 5]
```diff
@@ -0,0 +1,1375 @@
+/*
```
[the remark this answers is not in the document]
Comment (nxf58150) [reply hostapd/hostapd_cli_cmds.c:1]: I created new file hostap_cli_cmds.c due to below reasons:
1. There is `main` function in hostapd_cli.c and we don't need it on Zephyr. 
2. On Zephyr, the hostapd_cli isn't used with interactive mode.
3. We need to use global parameter `hostapd_cli_commands` in Zephyr SDK when doing hostapd initialization. However, in hostapd_cli.c, this global parameter is defined as "static". In hostapd_cli_cmds.c, I removed the static keyword for this parameter.

File: hostapd/ctrl_iface_zephyr.c [hunk 6]
```diff
@@ -0,0 +1,64 @@
+/*
+ * Hostapd / Zephyr socket pair -based control interface
+ * Copyright (c) 2022, Jouni Malinen <j@w1.fi>
```
[the remark this answers is not in the document]
Comment (nxf58150) [reply hostapd/ctrl_iface_zephyr.c:3]: Added NXP copyright here.

File: hostapd/ctrl_iface_zephyr.h [hunk 7]
```diff
@@ -0,0 +1,28 @@
+/*
+ * Hostapd / Zephyr socket pair -based control interface
+ * Copyright (c) 2022, Jouni Malinen <j@w1.fi>
```
[the remark this answers is not in the document]
Comment (nxf58150) [reply hostapd/ctrl_iface_zephyr.h:3]: Added NXP copyright here.

File: hostapd/ctrl_iface_zephyr.c [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+		send(sock, reply, reply_len, 0);
+	} else if (reply_len == 0) {
+		send(sock, "OK\n", 3, 0);
+	} else if (reply_len < 0) {
+		send(sock, "FAIL\n", 5, 0);
+	}
+	os_free(reply);
+}
+
+
```
[the remark this answers is not in the document]
Comment (nxf58150) [reply hostapd/ctrl_iface_zephyr.c]: Removed.

File: hostapd/ctrl_iface_zephyr.h [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+#define MAX_CTRL_MSG_LEN 1024
+
+struct conn_msg {
+	int msg_len;
+	char msg[MAX_CTRL_MSG_LEN];
+};
+
+void hostapd_ctrl_iface_receive(int sock, void *eloop_ctx, void *sock_ctx);
+
```
[the remark this answers is not in the document]
Comment (nxf58150) [reply hostapd/ctrl_iface_zephyr.h]: Removed.

File: hostapd/hostapd_cli_zephyr.h (hunk 1 above)
Comment (nxf58150) [reply hostapd/hostapd_cli_zephyr.h]: Updated.

File: src/drivers/driver_zephyr.c (hunk 2 above)
Comment (nxf58150) [reply src/drivers/driver_zephyr.c]: Updated.

File: src/drivers/driver_zephyr.c (hunk 3 above)
Comment (nxf58150) [reply src/drivers/driver_zephyr.c]: Those comment out callback functions might be supported in the future. I already removed commented lines to keep code clean. Thanks!

File: src/drivers/driver_zephyr.c (hunk 4 above)
Comment (nxf58150) [reply src/drivers/driver_zephyr.c:913]: Added commit message for each commit. Thanks!

File: src/drivers/driver_zephyr.c [hunk 10]
```diff
@@ -1102,6 +1181,121 @@ static void wpa_drv_zep_deinit(void *priv)
...
+{
+	struct zep_drv_if_ctx *if_ctx              = NULL;
+	const struct zep_wpa_supp_dev_ops *dev_ops = NULL;
+	const struct device *device                = NULL;
+	const struct net_linkaddr *link_addr       = NULL;
+	struct zep_hostapd_dev_callbk_fns callbk_fns;
+
+	device = net_if_get_device(net_if_get_wifi_uap());
+	if (!device) {
+		wpa_printf(MSG_ERROR, "%s: Interface %s not found", __func__, params->ifname);
```
Comment (jukkar) [inline src/drivers/driver_zephyr.c:1223]: This looks weird, shouldn't we try to figure out the device or net_if from the `params->ifname`?
We can get the net_if using `net_if_get_wifi_uap()`, but shouldn't we then also verify that the `params->ifname` is the same as what the net_if is? 
For example something like this:
```
if (net_if_get_wifi_uap() != net_if_get_by_index(net_if_get_by_name(params->ifname))) {
...interface not found...
}
```

File: hostapd/ctrl_iface.c [hunk 11]
```diff
@@ -3567,6 +3578,7 @@ static int hostapd_ctrl_iface_receive_process(struct hostapd_data *hapd,
 	} else if (os_strncmp(buf, "STA-NEXT ", 9) == 0) {
 		reply_len = hostapd_ctrl_iface_sta_next(hapd, buf + 9, reply,
 							reply_size);
+#ifndef __ZEPHYR__
```
Comment (krish2718) [inline hostapd/ctrl_iface.c:3581]: These are needed for monitoring internal events and are already supported for WPA supplicant, but if you want can add a TODO so that at least we are aware of the pending item?

File: hostapd/hostapd_cli_cmds.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (krish2718) [reply hostapd/hostapd_cli_cmds.c:1]: We had followed a similar approach for `wpa_cli.c`, but that file isn't included at all, so, had to carve out commands to a common file that can be used `wpa_cli.c` and `wpa_cli_zephyr.c`.

File: src/drivers/driver_zephyr.c [hunk 12]
```diff
@@ -1997,6 +2352,35 @@ int wpa_drv_zep_send_mlme(void *priv, const u8 *data, size_t data_len, int noack
 	return ret;
 }
 
+#ifdef CONFIG_WIFI_NM_HOSTAPD_AP
+int wpa_drv_hapd_send_eapol(void *priv, const u8 *addr, const u8 *data, size_t data_len,
```
Comment (krish2718) [inline src/drivers/driver_zephyr.c:2377]: If there are no changes b/w the two variants, do we need the `#ifdef`, can't we reuse the same function?

File: hostapd/ctrl_iface.c (hunk 11 above)
Comment (nxf58150) [reply hostapd/ctrl_iface.c:3582]: Added TODO to track this pending item.

File: src/drivers/driver_zephyr.c (hunk 12 above)
Comment (nxf58150) [reply src/drivers/driver_zephyr.c:2376]: Updated. Moved `#ifdef` into the function since data structures are different between wpa_supplicant and hostapd.

File: hostapd/hostapd_cli_cmds.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (nxf58150) [reply hostapd/hostapd_cli_cmds.c:1]: Updated file hostapd_cli_cmds.c

File: src/drivers/driver_zephyr.c (hunk 10 above)
Comment (nxf58150) [reply src/drivers/driver_zephyr.c]: Updated.

File: hostapd/hostapd_cli_cmds.c [hunk 13]
```diff
@@ -0,0 +1,1374 @@
...
+static int hostapd_cli_cmd_new_sta(struct wpa_ctrl *ctrl, int argc,
+				   char *argv[])
+{
+	char buf[64];
+	if (argc != 1) {
+		printf("Invalid 'new_sta' command - exactly one argument, STA "
+		       "address, is required.\n");
+		return -1;
+	}
+	snprintf(buf, sizeof(buf), "NEW_STA %s", argv[0]);
```
Comment (jukkar) [inline hostapd/hostapd_cli_cmds.c:207]: Perhaps we should have here `os_snprintf()` like in other functions.

File: hostapd/hostapd_cli_cmds.c [hunk 14]
```diff
@@ -0,0 +1,1374 @@
...
+static int hostapd_cli_cmd_sa_query(struct wpa_ctrl *ctrl, int argc,
+				    char *argv[])
+{
+	char buf[64];
+	if (argc != 1) {
+		printf("Invalid 'sa_query' command - exactly one argument, "
+		       "STA address, is required.\n");
+		return -1;
+	}
+	snprintf(buf, sizeof(buf), "SA_QUERY %s", argv[0]);
```
Comment (jukkar) [inline hostapd/hostapd_cli_cmds.c:299]: same here. Could you check the whole file and change accordingly.

**Review by jukkar — COMMENTED** [review]
Couple of minor nits. @krish2718 could you give this a look, this is the oldest pending PR so we could try to get it merged first.

File: hostapd/hostapd_cli_cmds.c (hunk 13 above)
Comment (nxf58150) [reply hostapd/hostapd_cli_cmds.c]: Updated.

File: hostapd/hostapd_cli_cmds.c [hunk 15]
```diff
@@ -173,9 +173,9 @@ static int hostapd_cli_cmd_sta(struct wpa_ctrl *ctrl, int argc, char *argv[])
 		return -1;
 	}
 	if (argc > 1)
-		snprintf(buf, sizeof(buf), "STA %s %s", argv[0], argv[1]);
+		os_snprintf(buf, sizeof(buf), "STA %s %s", argv[0], argv[1]);
```
Comment (krish2718) [inline hostapd/hostapd_cli_cmds.c:176]: can you please move this rename to a separate commit, it's not matching the commit description.

File: src/drivers/driver_zephyr.h [hunk 16]
```diff
@@ -200,7 +200,6 @@ struct zep_wpa_supp_dev_callbk_fns {
 			      union wpa_event_data *event);
 };
 
-#ifdef CONFIG_WIFI_NM_HOSTAPD_AP
```
Comment (krish2718) [inline src/drivers/driver_zephyr.h]: This a fixup and fixup commit aren't allowed in a PR, so, please use `fixup` and merge to the original commit,

**Comment by krish2718** [comment]
@nxf58150  can you please raise a manifest PR for this?

File: hostapd/hostapd_cli_cmds.c (hunk 15 above)
Comment (nxf58150) [reply hostapd/hostapd_cli_cmds.c:176]: I move this rename to the origin commit(2nd commit), which added hostapd_cli_cmds.c as new file. Is it ok for you?

File: src/drivers/driver_zephyr.h (hunk 16 above)
Comment (nxf58150) [reply src/drivers/driver_zephyr.h]: Updated.

**Comment by nxf58150** [comment]
> @nxf58150 can you please raise a manifest PR for this?

@krish2718 Do you want me to raise a PR on zephyr repo and hook this PR in the manifest file?

**Comment by krish2718** [comment]
> > @nxf58150 can you please raise a manifest PR for this?
> 
> @krish2718 Do you want me to raise a PR on zephyr repo and hook this PR in the manifest file?

Yes, for every PR to any repo in `west.yml` a.k.a, `west project`, there should be a corresponding PR to `zephyr` that updates `west.yml` a.k.a `mainfest PR` with `pull/<ID>/head` and once it's merged update with specific commit.

**Comment by nxf58150** [comment]
> > > @nxf58150 can you please raise a manifest PR for this?
> > 
> > 
> > @krish2718 Do you want me to raise a PR on zephyr repo and hook this PR in the manifest file?
> 
> Yes, for every PR to any repo in `west.yml` a.k.a, `west project`, there should be a corresponding PR to `zephyr` that updates `west.yml` a.k.a `mainfest PR` with `pull/<ID>/head` and once it's merged update with specific commit.

Hi @krish2718 I created new PR to update hostap in west.yml. Please help review it. Thanks!
https://github.com/zephyrproject-rtos/zephyr/pull/76414

**Comment by nxf58150** [comment]
> > > > @nxf58150 can you please raise a manifest PR for this?
> > > 
> > > 
> > > @krish2718 Do you want me to raise a PR on zephyr repo and hook this PR in the manifest file?
> > 
> > 
> > Yes, for every PR to any repo in `west.yml` a.k.a, `west project`, there should be a corresponding PR to `zephyr` that updates `west.yml` a.k.a `mainfest PR` with `pull/<ID>/head` and once it's merged update with specific commit.
> 
> Hi @krish2718 I created new PR to update hostap in west.yml. Please help review it. Thanks! [zephyrproject-rtos/zephyr#76414](https://github.com/zephyrproject-rtos/zephyr/pull/76414)

@krish2718 @jukkar 
I moved the manifest change to PR #75224 as suggested by Jukkar. The above PR is already closed. Thanks!

File: hostapd/hostapd_cli_cmds.c [hunk 17]
```diff
@@ -0,0 +1,1374 @@
...
+				      char *argv[])
+{
+	return hostapd_ctrl_command(ctrl, "GET_CONFIG");
+}
+
+
+static int wpa_ctrl_command_sta(struct wpa_ctrl *ctrl, const char *cmd,
+				char *addr, size_t addr_len, int print)
+{
+	char buf[4096], *pos;
```
Comment (jukkar) [inline hostapd/hostapd_cli_cmds.c:610]: Here too the buffer is quite large, is that really needed?

**Review by jukkar — COMMENTED** [review]
What is the status of this, there has not been much activity in three weeks so is this ready for new round of reviews?

**Comment by nxf58150** [comment]
> What is the status of this, there has not been much activity in three weeks so is this ready for new round of reviews?

We are still finding solutions to the hostapd event handler rename. Since hostap community didn't agree to apply my patch, I need to make my patch more generally. Chaitanya gave us a suggestion and we are still working on this.

File: hostapd/hostapd_cli_cmds.c [hunk 18]
```diff
@@ -0,0 +1,1374 @@
...
+#include "common/version.h"
+#include "common/cli.h"
+#include "hostapd_cli_zephyr.h"
+
+static DEFINE_DL_LIST(stations); /* struct cli_txt_entry */
+
+static int hostapd_cli_cmd(struct wpa_ctrl *ctrl, const char *cmd,
+			   int min_args, int argc, char *argv[])
+{
+	char buf[4096];
```
[the remark this answers is not in the document]
Comment (nxf58150) [reply hostapd/hostapd_cli_cmds.c]: Updated it to 1024.

File: hostapd/hostapd_cli_cmds.c (hunk 17 above)
Comment (nxf58150) [reply hostapd/hostapd_cli_cmds.c]: Updated it to 1024.

**Comment by nxf58150** [comment]
@krish2718 @jukkar 
Hi Chaitanya and Jukkar,
All the comments are replied and updated. For the rename change, we will find a proper solution in the future. Is this PR ok for you to merge? 
Thanks!

**Review by jukkar — APPROVED** [review]

**Comment by gannicus-gj** [comment]
Hi @krish2718 
could you pls vote for this PR? jukkar has approved, thanks

**Review by krish2718 — COMMENTED** [review]
first (oldest) commit should also now be `noup` as it was rejected by upstream community.

**Comment by nxf58150** [comment]
> first (oldest) commit should also now be `noup` as it was rejected by upstream community.

Updated to [noup], thanks!

**Review by krish2718 — APPROVED** [review]

**Comment by nxf58150** [comment]
@jukkar @krish2718 
Hi Jukkar and Chaitanya,
I saw both of you approved this PR. When can this PR be merged?
Thanks!

**Comment by krish2718** [comment]
> @jukkar @krish2718 Hi Jukkar and Chaitanya, I saw both of you approved this PR. When can this PR be merged? Thanks!

once the manifest PR CI is passed.
